privacyidea/lib/apps.py

```diff
@@ -56,19 +57,25 @@
 
 def _construct_extra_parameters(extra_data):
     """
-    Given a dictionary of extra key-value pairs (all unicode strings),
+    Given a dictionary of extra key-value pairs (all str),
     return a string that may be appended to a google authenticator / oathtoken URL.
-    Keys and values are converted to strings and urlquoted. Unicodes are converted to UTF-8.
-    :return: a string (may be empty if ``extra_data`` is empty
+    Values that are non-strings will be converted to str.
+    Keys and values are converted to UTF-8 and urlquoted.
+    :return: a string (may be empty if ``extra_data`` is empty)
     """
     extra_data_list = []
     for key, value in extra_data.items():
-        if isinstance(key, unicode):
+        # ``value`` might not be a str, in which case we will convert it.
+        if not isinstance(value, six.text_type):
+            value = str(value)
+        # We ensure ``key`` and ``value`` are bytes on Python 2 and 3.
+        if isinstance(key, six.text_type):
             key = key.encode('utf-8')
-        if isinstance(value, unicode):
+        if isinstance(value, six.text_type):
             value = value.encode('utf-8')
-        extra_data_list.append('{key}={value}'.format(key=quote(str(key)),
-                                                      value=quote(str(value))))
+        # Thus, ``quote`` returns (ASCII) bytes on Python 2. On Python 3, it returns str.
+        extra_data_list.append('{key}={value}'.format(key=quote(key),
+                                                      value=quote(value)))
```

Comment:
i was thinking about this section as well and thought about a helper function like `to_unicode_string(s)` which
 * does nothing if `s` already is a unicode string
 * converts `s` to a unicode string if it is a byte string
 * converts `s` to a (unicode) string if it is something else (int, bool, ...)
Similar to `convert_column_to_unicode()`
What do You think?

Reply:
This section is really a bit hacky :-) Maybe a helper function would also improve readability. But I'm not sure if a ``to_unicode_string`` function would be helpful here: ``key`` and ``value`` actually need to be converted to byte strings, because ``quote()`` on Python 2 crashes if we pass a unicode string:
```python
>>> urllib2.quote(u'ö')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python2.7/urllib.py", line 1298, in quote
    return ''.join(map(quoter, s))
KeyError: u'\xf6'
```

Reply:
I ran into this myself. That is why i added the [`to_byte_string()`](https://github.com/privacyidea/privacyidea/blob/911ac0da6baba09e5677cae6790cd8e66f027c84/privacyidea/lib/utils.py#L179) function.

Reply:
``_construct_extra_parameters`` now uses ``to_byte_string`` and it's much shorter and, I think, readable :-)